runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java

```diff
@@ -736,6 +737,11 @@ private void translateStreamingImpulse(
                 valueCoder.getClass().getSimpleName()));
       }
       keyCoder = ((KvCoder) valueCoder).getKeyCoder();
+      if (keyCoder instanceof LengthPrefixCoder) {
+        // Remove any unnecessary length prefixes which add more payload
+        // but also are not expected for state requests inside the operator.
+        keyCoder = ((LengthPrefixCoder) keyCoder).getValueCoder();
```

Comment:
The problem is that the `keyCoder` (extracted from the `KvCoder` above) is wrapped with a `LengthPrefixCoder`. This adds an unnecessary length prefix to the key and affects the partitioning of the keys. The state requests are served without the length prefix which requires us to either re-add it or remove it entirely (like we do here).